### PR TITLE
I1383 - switch from alt.js to redux for auth in webapps

### PR DESCRIPTION
**Pull request by georgearrowwood**

### CODE REVIEW

File: app/src/test/report/components/ReportingAppTests.tsx [hunk 1]
```diff
@@ -15,13 +15,13 @@ describe("ReportingApp", () => {
                 errors: ["a", "b"],
                 infos: ["c", "d"]
             },
-            ReportingAuthStore: { loggedIn: true }
+            // ReportingAuthStore: { loggedIn: true }
```
Comment (littledice) [inline app/src/test/report/components/ReportingAppTests.tsx]: can remove commented code

File: app/src/test/shared/components/Login/LoggedInUserBoxTests.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            loggedIn={true}
+            username="test.user"
+            logOut={()=>({})}
+        />);
+        expect(rendered.text()).to.contain("Logged in as test.user");
+        expect(rendered.find(InternalLink)).to.have.length(1);
+    });
+
+    it("clicking log out emits logOut event", () => {
+        const dispatchSpy = sandbox.createSpy();
```
Comment (littledice) [inline app/src/test/shared/components/Login/LoggedInUserBoxTests.tsx]: This test checks that clicking logout triggers the prop function `logOut`, but really we'd like to test that the `logOut` function is mapped to the right thing by `mapStateToProps`, and that `authActions.logOut()` is called.

File: app/src/test/shared/stores/NotificationStoreTests.ts [hunk 3]
```diff
@@ -88,39 +88,39 @@ describe("NotificationStore", () => {
...
-            sub: "user",
-            permissions: "*/can-login",
-            roles: ""
-        }, "secret");
-        authActions.logIn(token, false);
-
-        const state = notificationStore.getState();
-        expect(state.errors[0]).to.contain("Only members of modelling groups");
-    });
+    // it("logIn does not set errorMessage if user is active modeller", () => {
```
Comment (littledice) [inline app/src/test/shared/stores/NotificationStoreTests.ts]: Is the functionality in these tests still implemented and tested somewhere?

File: app/src/test/shared/fetch/helpers.ts [hunk 4]
```diff
@@ -137,26 +136,26 @@ export class FetchHelper<TPayload, TActionPayload> {
...
-                errorMessage: null,
-                expectedAction: {
-                    action: "AuthActions.logOut",
-                    payload: true
-                },
-                triggerFetch: this.config.triggerFetch
-            });
-        });
+
+        // it("triggers logout when server returns bearer token error", (done: DoneCallback) => {
```
Comment (littledice) [inline app/src/test/shared/fetch/helpers.ts]: This functionality should still exist (and be tested!) and also I think needs to be tested with respect to the new `LocalService` class - i.e. we want to make sure that the user is logged out when the service gets a `token expired` error back from the API

File: app/src/main/shared/sources/Source.ts [hunk 5]
```diff
@@ -95,7 +96,10 @@ function processResult<TModel>(result: Result, response: any): TModel | void {
         switch (error.code) {
             case "bearer-token-invalid":
                 console.log("Access token has expired or is otherwise invalid: Logging out.");
-                authActions.logOut();
+                // Temporary log off
+                localStorageHandler.remove("accessToken");
+                location.reload();
```
Comment (littledice) [inline app/src/main/shared/sources/Source.ts]: :+1:

**Comment by littledice** [comment]
These tests look a lot more comprehensive now - awesome :)

File: app/src/test/shared/components/Login/LoggedInUserBoxTests.tsx (hunk 2 above)
Comment (littledice) [reply app/src/test/shared/components/Login/LoggedInUserBoxTests.tsx]: Note to self - this is now tested below :+1:

File: app/src/test/shared/fetch/helpers.ts (hunk 4 above)
Comment (littledice) [reply app/src/test/shared/fetch/helpers.ts]: As discussed, the old fetcher is now just removing the token from local storage and refreshing the page on a bearer token error, which has the effect of logging the user out. It's a temporary hack and the fetcher will be phased out. The functionality of the new `LocalService` is all under test. So we can go ahead and delete this test!

Comment (littledice) [reply app/src/test/shared/fetch/helpers.ts]: @georgearrowwood  nice job, I think we can delete this commented code now ^

File: app/webpack.config.js [hunk 6]
```diff
@@ -25,6 +25,7 @@ function makePortalConfig(name, urlPrefix) {
         name: name,
 
         entry: [
+        	"babel-regenerator-runtime",
```
Comment (MartinEden) [inline app/webpack.config.js]: What does this do?

Also, could you please fix indentation here.

File: app/src/test/contrib/components/LogoutComponentTests.tsx [hunk 7]
```diff
@@ -1,42 +0,0 @@
...
-        const state = Object.assign(initialAuthState(), {
-            loggedIn: true,
-            username: "test.user"
-        });
-        const rendered = shallow(<LogoutComponent { ...state } />);
-        const spy = sandbox.dispatchSpy();
-        rendered.find(InternalLink).simulate("click");
-        expectOneAction(spy, { action: "AuthActions.logOut" });
-    });
-});
```
Comment (MartinEden) [inline app/src/test/contrib/components/LogoutComponentTests.tsx]: Are there tests to replace these?

File: app/src/test/shared/actions/AuthActionsTests.ts [hunk 8]
```diff
@@ -1,49 +1,126 @@
...
+    afterEach(() => {
         sandbox.restore();
-    }
-}
+    });
 
-describe("AuthActions", () => {
-    it("retrieves username from token", () => {
-        const payload = getPayload({ sub: "username", permissions: "", roles: "" });
-        expect(payload.username).to.equal("username");
```
Comment (MartinEden) [inline app/src/test/shared/actions/AuthActionsTests.ts]: Are there tests somewhere that show tokens get properly decoded?

File: app/src/test/shared/components/LoginFormTests.tsx [hunk 9]
```diff
@@ -1,132 +0,0 @@
...
-        });
-        checkSubmit(form, done, sandbox, spy => {
-            actionHelpers.expectOrderedActions(
-                spy,
-                [{ action: "Login_test/submitFailed", payload: "An error occurred logging in" }],
-                numberOfSubmissionActions
-            );
-        });
-    });
-});
```
Comment (MartinEden) [inline app/src/test/shared/components/LoginFormTests.tsx]: Are there tests for the new log in form somewhere?

File: app/src/main/contrib/actions/modellingGroupsActions.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { Dispatch } from "redux";
+
+import { ModellingGroupsService } from "../services/ModellingGroupsService";
+import {GroupsFetched, ModellingGroupTypeKeys} from "../actionTypes/ModellingGroupsTypes";
+
+export const modellingGroupsActions = {
+
+    getGroups() {
+        return async (dispatch: Dispatch<any>, getState: Function) => {
```
[the remark this answers is not in the document]
Comment (georgearrowwood) [reply app/src/main/contrib/actions/modellingGroupsActions.ts]: Could not find the way to type getState, tried to google it, people mostly just use Function as type for it, so I did

File: app/src/main/shared/components/Login/LoginForm.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+import { InternalLink } from "../InternalLink";
+import { ValidationError } from "./ValidationError";
+import { validations } from "../../modules/ReduxForm";
+import { authActions } from "../../actions/authActions";
+import { GlobalState } from "../../reducers/GlobalState";
+import {AuthActionsTypes} from "../../actionTypes/AuthTypes";
+
+export interface LoginFormProps {
+    handleSubmit: (F: any) => any;
```
[the remark this answers is not in the document]
Comment (georgearrowwood) [reply app/src/main/shared/components/Login/LoginForm.tsx]: handleSubmit is internal function of redux-form that is running given callback function after submit

File: app/src/main/shared/services/LocalService.ts [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+    protected getTokenFromState(state: any) {
+        if (state.auth && state.auth.bearerToken) {
+            return state.auth.bearerToken
+        }
+    }
+
+    public setOptions(options: any) {
+        Object.assign(this.options, options);
+    }
```
[the remark this answers is not in the document]
Comment (georgearrowwood) [reply app/src/main/shared/services/LocalService.ts]: here i merge options generated in initOptions with given options

File: app/webpack.config.js (hunk 6 above)
Comment (georgearrowwood) [reply app/webpack.config.js]: this is module which allows to use async await functionality

File: app/src/test/contrib/components/LogoutComponentTests.tsx (hunk 7 above)
Comment (georgearrowwood) [reply app/src/test/contrib/components/LogoutComponentTests.tsx]: Logout tests are here:
montagu-webapps/app/src/test/shared/components/Login/LoggedInUserBoxTests.tsx

File: app/src/test/shared/components/LoginFormTests.tsx (hunk 9 above)
Comment (georgearrowwood) [reply app/src/test/shared/components/LoginFormTests.tsx]: montagu-webapps/app/src/test/shared/components/Login/LoginFormTests.tsx

File: app/src/test/shared/actions/AuthActionsTests.ts (hunk 8 above)
Comment (georgearrowwood) [reply app/src/test/shared/actions/AuthActionsTests.ts]: montagu-webapps/app/src/test/shared/modules/JwtTokenTests.ts

**Pull request merged**